### PR TITLE
Define queries using a proc macro

### PR DESCRIPTION
**Pull request by Zoxc**
cc @rust-lang/compiler

### CODE REVIEW

File: src/librustc/ty/query/mod.rs [hunk 1]
```diff
@@ -103,11 +103,9 @@ pub use self::on_disk_cache::OnDiskCache;
 // (error) value if the query resulted in a query cycle.
 // Queries marked with `fatal_cycle` do not need the latter implementation,
 // as they will raise an fatal error on query cycles instead.
-define_queries! { <'tcx>
-    Other {
-        /// Records the type of every item.
-        [] fn type_of: TypeOfItem(DefId) -> Ty<'tcx>,
 
+rustc_query_append! { [define_queries!][ <'tcx>
```
Comment (eddyb) [inline src/librustc/ty/query/mod.rs]: This is mind-bending, but cool!

File: src/librustc/query/mod.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// Each of these quries corresponds to a function pointer field in the
```
Comment (estebank) [inline src/librustc/query/mod.rs]:
```suggestion
// Each of these queries corresponds to a function pointer field in the
```

**Comment by mark-i-m** [comment]
@Zoxc This is cool :)

When it merges, could you please update https://rust-lang.github.io/rustc-guide/query.html ? That would be much appreciated!

File: src/librustc_macros/src/query.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        let content;
+        braced!(content in input);
+        Ok(Group {
+            name,
+            queries: content.parse()?,
+        })
+    }
+}
+
+fn camel_case(string: &str) -> String {
```
Comment (eddyb) [inline src/librustc_macros/src/query.rs]: I don't think we should do this. We already use the `snake_case` query name for `struct` / `enum` variant names, it's just dep nodes that are different because they started outside of the query system.

One (significant?) advantage of not changing the case of query names, at all, is greppability: `type_of` is `type_of` *everywhere* the query is mentioned, not `TypeOf` in some places.

cc @nikomatsakis

**Comment by eddyb** [comment]
@petrochenkov This is pretty much just replacing a `macro_rules` macro (`define_queries!`) with a proc macro that has nicer (and simpler) input syntax.

The PR is more confusing atm though, because it kept both macros and daisy-chained them, to allow gradually moving from the old one to the new one.

File: src/librustc/ty/query/README.md [hunk 4]
```diff
@@ -218,7 +218,7 @@ define_queries! { <'tcx>
 Each line of the macro defines one query. The name is broken up like this:
 
 ```
-[] fn type_of: TypeOfItem(DefId) -> Ty<'tcx>,
+[] fn type_of: TypeOf(DefId) -> Ty<'tcx>,
 ^^    ^^^^^^^  ^^^^^^^^^^ ^^^^^     ^^^^^^^^
```
Comment (eddyb) [inline src/librustc/ty/query/README.md]: These markers were not updated.

**Comment by Zoxc** [comment]
I forgot to include syntax for `QueryDescription`:
```rust

impl<'tcx> QueryDescription<'tcx> for queries::type_op_prove_predicate<'tcx> {
    fn describe(_tcx: TyCtxt<'_, '_, '_>, goal: CanonicalTypeOpProvePredicateGoal<'tcx>)
                -> Cow<'static, str> {
        format!("evaluating `type_op_prove_predicate` `{:?}`", goal).into()
    }
}
```
Anyone have some good ideas for that?

File: src/librustc_macros/src/query.rs (hunk 3 above)
Comment (Zoxc) [reply src/librustc_macros/src/query.rs]: That only really matters to incremental tests though. I'm not sure I actually want them to show up in search results =P

**Comment by eddyb** [comment]
I'd say something like this:
```rust
description("evaluating `type_op_prove_predicate` `{:?}`", goal)
```

**Comment by Zoxc** [comment]
@eddyb That looks very unbound.

How about:
```rust
query type_of(DefId) -> Ty<'tcx> {
 	"evaluating `finding type of `{:?}`", $key
}
```

**Comment by eddyb** [comment]
@mark-i-m Literally this?
```rust
/// "finding type of `{:?}`", $key`
```
Or something neater? Maybe like this:
```rust
/// finding type of `$key`
```

**Comment by Zoxc** [comment]
I don't think using comments is a good idea. You'd want to use those to document the queries. Also it's very surprising that comments are executable.

**Comment by Zoxc** [comment]
Maybe this?
```
query type_of(key: DefId) -> Ty<'tcx> {
	cache { key.is_local() },
	description { "evaluating `finding type of `{:?}`", key },
	fatal_cycle,
	eval_always,
}
```

**Comment by Zoxc** [comment]
cc @alexcrichton for the build system changes

**Comment by Zoxc** [comment]
@alexcrichton Well the build system changes there are not correct. Which is why I wanted you to look at the changes here. We can move them over to that PR and land that first though.

File: src/librustc/lib.rs [hunk 5]
```diff
@@ -86,6 +88,7 @@ extern crate graphviz;
 extern crate libc;
 extern crate polonius_engine;
 extern crate rustc_target;
+#[macro_use] extern crate rustc_macros;
```
Comment (alexcrichton) [inline src/librustc/lib.rs]: FWIW wherever possible it'd be great to stop using `#[macro_use]` for new macro-based code

**Review by alexcrichton — COMMENTED** [review]
Ah ok, makes sense! The build system changes here look good to me (or at least not obviously wrong). I still have basically no idea how this works without updating the bootstrap compiler! I'm going to check this out locally and poke around it a bit too.

**Comment by alexcrichton** [comment]
Ok I've sat down with some build logs and I think I've figured this out. There's basically a whole slew of decisions, none of which were made to enable this, which are enabling this.

First up, some facts:

* When compiling procedural macros,  [`RUSTC_SNAPSHOT` is always chosen](https://github.com/rust-lang/rust/blob/1137d29d5e551e377579c5a601fe7c444057d00c/src/bootstrap/bin/rustc.rs#L89). This is because there is no `--target` argument
* For the rustc component of building, this env var [points to the compiler in question](https://github.com/rust-lang/rust/blob/1137d29d5e551e377579c5a601fe7c444057d00c/src/bootstrap/builder.rs#L980-L982), and it's not always hardcoded to the actual stage0 rustc binary (like I thought). This `self.rustc(..)` and `self.rustc_libdir(..)` both have special cases for the stage0 case (snapshot) - [A](https://github.com/rust-lang/rust/blob/1137d29d5e551e377579c5a601fe7c444057d00c/src/bootstrap/builder.rs#L659-L668), [B](https://github.com/rust-lang/rust/blob/1137d29d5e551e377579c5a601fe7c444057d00c/src/bootstrap/builder.rs#L633-L644)

Next I want to explain what happens during the first compile. During the first compilation of the compiler the `rustc_macros` crate here must be linked against `proc_macro`. While building libstd/libtest we create a `libproc_macro.rlib` binary. This is entirely unused by `rustc_macros`! Instead due to the above we're using the stage0 (snapshot) compiler and linking against the `proc_macro` in the sysroot, **which is the stage0 sysroot that we downloaded**. This means that during the first compilation of the compiler, procedural macros are using the `proc_macro` crate from the original compiler, making it trivially compatible with the compiler as they're all compiled against it.

The second compilation is a bit different. We don't have any downloaded libraries to use and we're on our own. By the time we reach `rustc_macros` the `libproc_macro.rlib` file has again been compiled during libstd/libtest. This time, however, it's actually used when compiling `rustc_macro` as it's in the sysroot that we're using (as there's not two sysroots in stage1 unlike in stage0). The stage1 compiler then loads what it just compiled and runs it.

---

I believe this means that the first compilation of `libproc_macro` is basically inert. It could be full of panics and it doesn't matter. This also means that if we change the ABI of the structures in `proc_macro` it's ok. Only the stage1 compiler actually starts using it, so the stage1 compiler is expecting the new ABI and the `proc_macro` crate we build is providing the new ABI. 

Finally, just to justify to myself, we do still need @eddyb's work on a "stable ABI" for `proc_macro`. If we didn't have that, then the stage1 compiler would have a stage0-defined ABI, but it may be compiling a new ABI which it's loading. Put another way, the stage0 and stage1 compiler disagree about the "Rust" ABI, but they agree on the "C" ABI, which is what we need.

So... actually I think this is it? I don't think there's anything else needed here, and we can land this at any time irrespective of updating the bootstrap compiler. It's all set up quite precariously but it at least works!

File: src/bootstrap/bin/rustc.rs [hunk 6]
```diff
@@ -294,6 +287,13 @@ fn main() {
         }
     }
 
+    // Force all crates compiled by this compiler to (a) be unstable and (b)
```
Comment (bjorn3) [inline src/bootstrap/bin/rustc.rs]: Why is this moved?

File: src/librustc_macros/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+version = "0.1.0"
+authors = ["The Rust Project Developers"]
+
+[lib]
+proc-macro = true
+
+[dependencies]
+syn = { version = "0.15.22", features = ["full"] }
+proc-macro2 = "0.4.24"
+quote = "0.6.10"
```
Comment (bjorn3) [inline src/librustc_macros/Cargo.toml]: Missing trailing newline

**Comment by alexcrichton** [comment]
One thing I've just thought of is that we may want to start out using procedural macros in rustc with something that's minimally invasive to rustc itself (aka easy to revert). It may take a few days to ensure that we distribute the compiler correctly and various tooling like rls/rustfmt/clippy/etc continue to work, so I just want to make sure that if we need to fix a bug and it takes awhile for whatever reason we don't leave nightlies inert for some time

**Comment by Zoxc** [comment]
@alexcrichton Do we test cross compilation?

**Comment by alexcrichton** [comment]
We do sort of in the sense that we cross compile a lot of things, but we don't test cross compilation in the way that it looks to end users I think much. Or at least I think the ecosystem tests it a lot more than we do!

**Comment by TimNN** [comment]
Ping from triage! What is the status of this PR?

**Comment by Zoxc** [comment]
Still blocked on https://github.com/rust-lang/rust/pull/56795

**Comment by Zoxc** [comment]
This is now unblocked.

r? @michaelwoerister

File: src/librustc_macros/src/tt.rs [hunk 8]
```diff
@@ -0,0 +1,65 @@
+use proc_macro2::{Delimiter, TokenStream, TokenTree};
+
+pub struct TT<'a>(pub &'a TokenTree);
```
Comment (Zoxc) [inline src/librustc_macros/src/tt.rs]: This code was extracted from `syn`. Let's make sure the licensing works out.

Comment (oli-obk) [reply src/librustc_macros/src/tt.rs]: Can't we just use `syn` as a dependency here?

File: src/librustc_macros/src/query.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc_macros/src/query.rs]: I think we should either fully change it to keep `snake_case`, or keep the original behavior (so append `Item` to the name as to not change any test)

File: src/librustc_macros/src/query.rs [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                name.span());
+            let arg = &query.arg;
+            let key = &query.key.0;
+            let result_full = &query.result;
+            let result = match query.result {
+                ReturnType::Default => quote! { -> () },
+                _ => quote! { #result_full },
+            };
+
+            // Find out if we should cache the query on disk
```
Comment (oli-obk) [inline src/librustc_macros/src/query.rs]: This is the only comment explaining what all the parsing code is doing. I think some more docs and comments are warranted given the complexity of it.

File: src/librustc_macros/src/tt.rs (hunk 8 above)
Comment (Zoxc) [reply src/librustc_macros/src/tt.rs]: The code is private in `syn`.

File: src/librustc_macros/src/query.rs (hunk 3 above)
Comment (Zoxc) [reply src/librustc_macros/src/query.rs]: I've changed the code to preserve the names.

File: src/librustc_macros/src/tt.rs (hunk 8 above)
Comment (oli-obk) [reply src/librustc_macros/src/tt.rs]: That seems resolvable :laughing: 	Please attempt to make the code public in syn (or public behind a cargo feature that also turns on something unstable as to not stabilize this if making it public is undesirable right now).

Comment (Zoxc) [reply src/librustc_macros/src/tt.rs]: @dtolnay Do you know of a way to parse just doc comments? So I could get rid of this code?

File: src/librustc_macros/src/query.rs [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+    result: ReturnType,
+}
+
+fn check_attributes(attrs: Vec<Attribute>) {
+    for attr in attrs {
+        let path = attr.path;
+        let path = quote! { #path };
+        let path = TS(&path);
+
+        if path != TS(&quote! { doc }) {
```
Comment (dtolnay) [inline src/librustc_macros/src/query.rs]: I would write this as `!attr.path.is_ident("doc")`.

File: src/librustc_macros/src/query.rs [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+            let id = args.parse()?;
+            if !args.is_empty() {
+                panic!("unexpected tokens in arguments");
+            };
+            let block = input.parse()?;
+            Ok(QueryAttribute::LoadCached(tcx, id, block))
+        } else if attr == "fatal_cycle" {
+            Ok(QueryAttribute::FatalCycle)
+        } else {
+            panic!("unknown query modifier {}", attr)
```
Comment (dtolnay) [inline src/librustc_macros/src/query.rs]: Here (and elsewhere) I would prefer to see the parsing code return a syn::Error rather than panic. That way the error can be correctly spanned to show which part of a potentially large macro input was problematic.

File: src/librustc_macros/src/query.rs [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+            let tcx = if attr_content.peek(Token![|]) {
+                attr_content.parse::<Token![|]>()?;
+                let tcx = attr_content.parse()?;
+                attr_content.parse::<Token![|]>()?;
+                Some(tcx)
+            } else {
+                None
+            };
+            let desc = attr_content.parse_terminated(Expr::parse)?;
+            if !attr_content.is_empty() {
```
Comment (dtolnay) [inline src/librustc_macros/src/query.rs]: No need for this check, `parse_terminated` parses the entire input.

File: src/librustc_macros/src/query.rs [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+    Desc(Option<Ident>, Punctuated<Expr, Token![,]>),
+    Cache(Option<Ident>, Expr),
+    LoadCached(Ident, Ident, Block),
+    FatalCycle,
+}
+
+impl Parse for QueryAttribute {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let attr: Ident = input.parse()?;
+        if attr == "desc" {
```
Comment (dtolnay) [inline src/librustc_macros/src/query.rs]: Please put a comment inside each of these blocks showing an example input that would be parsed. It's hard to read off from the parsing code.

File: src/librustc_macros/src/query.rs [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+            let tcx = if attr_content.peek(Token![|]) {
+                attr_content.parse::<Token![|]>()?;
+                let tcx = attr_content.parse()?;
+                attr_content.parse::<Token![|]>()?;
+                Some(tcx)
+            } else {
+                None
+            };
+            let expr = attr_content.parse()?;
+            if !attr_content.is_empty() {
```
Comment (dtolnay) [inline src/librustc_macros/src/query.rs]: No need for this check -- we catch when tokens do not get parsed.

File: src/librustc_macros/src/query.rs [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+                panic!("unexpected tokens in block");
+            };
+            Ok(QueryAttribute::Cache(tcx, expr))
+        } else if attr == "load_cached" {
+            let args;
+            parenthesized!(args in input);
+            let tcx = args.parse()?;
+            args.parse::<Token![,]>()?;
+            let id = args.parse()?;
+            if !args.is_empty() {
```
Comment (dtolnay) [inline src/librustc_macros/src/query.rs]: No need for this check.

File: src/librustc_macros/src/tt.rs (hunk 8 above)
Comment (dtolnay) [reply src/librustc_macros/src/tt.rs]: Yes, I suggested an easier way at the place this is used.

File: src/librustc_macros/src/query.rs (hunk 9 above)
Comment (Zoxc) [reply src/librustc_macros/src/query.rs]: I added some more of them.

File: src/librustc_macros/src/query.rs [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+    syn::custom_keyword!(query);
+}
+
+/// Ident or a wildcard `_`.
+struct IdentOrWild(Ident);
+
+impl Parse for IdentOrWild {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        Ok(if input.peek(Token![_]) {
+            input.parse::<Token![_]>()?;
```
Comment (oli-obk) [inline src/librustc_macros/src/query.rs]:
```suggestion
            let underscore = input.parse::<Token![_]>()?;
```

File: src/librustc_macros/src/query.rs [hunk 17]
```diff
@@ -0,0 +1,343 @@
...
+            IdentOrWild(Ident::new("_", Span::call_site()))
```
Comment (oli-obk) [inline src/librustc_macros/src/query.rs]:
```suggestion
            IdentOrWild(Ident::new("_", underscore.spans[0]))
```

File: src/librustc_macros/src/query.rs [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+/// Ensures only doc comment attributes are used
+fn check_attributes(attrs: Vec<Attribute>) -> Result<()> {
+    for attr in attrs {
+        if !attr.path.is_ident("doc") {
+            return Err(Error::new(attr.span(), "attributes not supported on queries"));
+        }
+    }
+    Ok(())
```
Comment (oli-obk) [inline src/librustc_macros/src/query.rs]: Note: don't address this, it just tickled my code golfing nerve endings... and the iterator chain is just so much less clear than the current code I had to post it.

```rust
attrs
    .iter()
    .filter(|attr| !attr.path.is_ident("doc"))
    .map(|attr| Err(Error::new(attr.span(), "attributes not supported on queries")))
    .collect()
```

File: src/librustc_macros/src/query.rs [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+}
+
+/// A compiler query. `query ... { ... }`
+struct Query {
+    attrs: List<QueryModifier>,
```
Comment (oli-obk) [inline src/librustc_macros/src/query.rs]: All use sites of this field do a `.iter().find_map(..)` to an `Option` on this field. Maybe just split it out into `Option` fields on the `Query` type?

**Review by oli-obk — CHANGES_REQUESTED** [review]
just a nit and a potential refactoring

r=me with both addressed

File: src/librustc_macros/src/query.rs (hunk 19 above)
Comment (Zoxc) [reply src/librustc_macros/src/query.rs]: This is supposed to be an AST for queries, maybe I should add a new type describing the modifiers?

Comment (oli-obk) [reply src/librustc_macros/src/query.rs]: Yes, that's also good.

Comment (Zoxc) [reply src/librustc_macros/src/query.rs]: I added a commit doing that.

**Review by oli-obk — APPROVED** [review]
